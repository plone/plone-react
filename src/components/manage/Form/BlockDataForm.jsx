import { defineMessages } from 'react-intl';
import { InlineForm } from '@plone/volto/components';
import { withBlockSchemaEnhancer } from '@plone/volto/helpers';

const messages = defineMessages({
  variation: {
    id: 'Variation',
    defaultMessage: 'Variation',
  },
});

const BlockDataForm = withBlockSchemaEnhancer(InlineForm, {
  extensionName: 'variation',
  title: messages.variation, // this is the title
});

export default BlockDataForm;
