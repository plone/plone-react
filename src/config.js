/**
 * Config.
 * @module config
 */
import { defaults } from 'lodash';

export default defaults(
  {},
  {
    host: process.env.HOST,
    port: process.env.PORT,
    apiPath: process.env.API_PATH,
  },
  {
    host: 'localhost',
    port: '4300',
    apiPath: 'http://localhost:8081/db/web',
  },
);
