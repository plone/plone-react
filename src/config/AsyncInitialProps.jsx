// import { getBaseUrl } from '@plone/volto/helpers';
// import config from '@plone/volto/registry';
// import {
//   getBreadcrumbs,
//   getContent,
//   getNavigation,
//   getTypes,
//   getWorkflow,
// } from '@plone/volto/actions';

const AsyncInitialProps = [
  {
    route: {
      path: '/**',
    },
    asyncProps: [
      // {
      //   key: 'breadcrumbs',
      //   promise: ({ location, store: { dispatch } }) => {
      //     return (
      //       __SERVER__ &&
      //       dispatch(getBreadcrumbs(getBaseUrl(location.pathname)))
      //     );
      //   },
      // },
      // {
      //   key: 'content',
      //   promise: ({ location, store: { dispatch } }) =>
      //     __SERVER__ && dispatch(getContent(getBaseUrl(location.pathname))),
      // },
      // {
      //   key: 'navigation',
      //   promise: ({ location, store: { dispatch } }) =>
      //     __SERVER__ &&
      //     dispatch(
      //       getNavigation(
      //         getBaseUrl(location.pathname),
      //         config.settings.navDepth,
      //       ),
      //     ),
      // },
      // {
      //   key: 'types',
      //   promise: ({ location, store: { dispatch } }) =>
      //     __SERVER__ && dispatch(getTypes(getBaseUrl(location.pathname))),
      // },
      // {
      //   key: 'workflow',
      //   promise: ({ location, store: { dispatch } }) =>
      //     __SERVER__ && dispatch(getWorkflow(getBaseUrl(location.pathname))),
      // },
    ],
  },
];

export default AsyncInitialProps;
