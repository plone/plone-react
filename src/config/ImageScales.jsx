export const imageScales = {
  large: 768,
  preview: 400,
  mini: 200,
  thumb: 128,
  tile: 64,
  icon: 32,
  listing: 16,
};
