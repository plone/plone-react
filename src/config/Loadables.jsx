import loadable from '@loadable/component';

export const loadables = {
  prettierStandalone: loadable.lib(() => import('prettier/standalone')),
  prettierParserHtml: loadable.lib(() => import('prettier/parser-html')),
  prismCore: loadable.lib(() => import('prismjs/components/prism-core')),
  toastify: loadable.lib(() => import('react-toastify'), {
    resolveComponent: (lib) => lib.toast,
  }),
  reactSelect: loadable.lib(() => import('react-select')),
};
