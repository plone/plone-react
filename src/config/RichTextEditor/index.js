import ToHTMLRenderers, { options as ToHTMLOptions } from './ToHTML';
import { extendedBlockRenderMap, blockStyleFn, listBlockTypes } from './Blocks';
import plugins, { inlineToolbarButtons } from './Plugins';
import FromHTMLCustomBlockFn from './FromHTML';

export {
  extendedBlockRenderMap,
  blockStyleFn,
  listBlockTypes,
  FromHTMLCustomBlockFn,
  inlineToolbarButtons as richTextEditorInlineToolbarButtons,
  plugins as richTextEditorPlugins,
  ToHTMLRenderers,
  ToHTMLOptions,
};
