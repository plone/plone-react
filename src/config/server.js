import imagesMiddleware from '@plone/volto/express-middleware/images';
import filesMiddleware from '@plone/volto/express-middleware/files';
import robotstxtMiddleware from '@plone/volto/express-middleware/robotstxt';
import sitemapMiddleware from '@plone/volto/express-middleware/sitemap';
import { readCriticalCss } from '@plone/volto/critical-css';

const settings = {
  expressMiddleware: [
    filesMiddleware(),
    imagesMiddleware(),
    robotstxtMiddleware(),
    sitemapMiddleware(),
  ],
  readCriticalCss,
};

export default settings;
