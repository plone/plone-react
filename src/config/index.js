/**
 * Config.
 * @module config
 */

import { defaultWidget, widgetMapping } from './Widgets';
import {
  layoutViews,
  contentTypesViews,
  defaultView,
  errorViews,
} from './Views';
import { nonContentRoutes } from './NonContentRoutes';
import ToHTMLRenderers, {
  options as ToHTMLOptions,
} from './RichTextEditor/ToHTML';
import {
  extendedBlockRenderMap,
  blockStyleFn,
  listBlockTypes,
} from './RichTextEditor/Blocks';
import plugins, { inlineToolbarButtons } from './RichTextEditor/Plugins';
import FromHTMLCustomBlockFn from './RichTextEditor/FromHTML';
import {
  groupBlocksOrder,
  requiredBlocks,
  blocksConfig,
  initialBlocks,
} from './Blocks';
import { loadables } from './Loadables';

import { sentryOptions } from './Sentry';
import { contentIcons } from './ContentIcons';

import applyAddonConfiguration from 'load-volto-addons';

const host = process.env.HOST || 'localhost';
const port = process.env.PORT || '3000';

const apiPath =
  process.env.RAZZLE_API_PATH ||
  (__DEVELOPMENT__
    ? `http://${host}:${port}/api`
    : 'http://localhost:8080/Plone');

let config = {
  settings: {
    host,
    port,
    // Internal proxy to bypass CORS *while developing*. Not intended for production use.
    // In production, the proxy is disabled, make sure you specify an apiPath that does
    // not require CORS to work.
    apiPath,
    devProxyToApiPath:
      process.env.RAZZLE_DEV_PROXY_API_PATH || 'http://localhost:8080/Plone', // Set it to '' for disabling the proxy
    // proxyRewriteTarget Set it for set a custom target for the proxy or overide the internal VHM rewrite
    // proxyRewriteTarget: '/VirtualHostBase/http/localhost:8080/Plone/VirtualHostRoot/_vh_api'
    // proxyRewriteTarget: 'https://myvoltositeinproduction.com'
    proxyRewriteTarget: process.env.RAZZLE_PROXY_REWRITE_TARGET || undefined,
    // apiPath: process.env.RAZZLE_API_PATH || 'http://localhost:8000', // for Volto reference
    // apiPath: process.env.RAZZLE_API_PATH || 'http://localhost:8081/db/web', // for guillotina
    actions_raising_api_errors: ['GET_CONTENT', 'UPDATE_CONTENT'],
    internalApiPath: process.env.RAZZLE_INTERNAL_API_PATH || undefined,
    websockets: process.env.RAZZLE_WEBSOCKETS || false,
    nonContentRoutes,
    extendedBlockRenderMap,
    blockStyleFn,
    listBlockTypes,
    FromHTMLCustomBlockFn,
    richTextEditorInlineToolbarButtons: inlineToolbarButtons,
    richTextEditorPlugins: plugins,
    ToHTMLRenderers,
    ToHTMLOptions,
    imageObjects: ['Image'],
    listingPreviewImageField: 'image',
    customStyleMap: null,
    notSupportedBrowsers: ['ie'],
    defaultPageSize: 25,
    isMultilingual: false,
    supportedLanguages: ['en'],
    defaultLanguage: 'en',
    navDepth: 1,
    expressMiddleware: [],
    defaultBlockType: 'text',
    verticalFormTabs: false,
    persistentReducers: ['blocksClipboard'],
    sentryOptions: {
      ...sentryOptions,
    },
    contentIcons: contentIcons,
    loadables,
    appExtras: [],
    maxResponseSize: 2000000000, // This is superagent default (200 mb)
  },
  widgets: {
    ...widgetMapping,
    default: defaultWidget,
  },
  views: {
    layoutViews,
    contentTypesViews,
    defaultView,
    errorViews,
  },
  blocks: {
    requiredBlocks,
    blocksConfig,
    groupBlocksOrder,
    initialBlocks,
  },

  addonRoutes: [],
  addonReducers: {},
};

config = applyAddonConfiguration(config);

export const settings = config.settings;
export const widgets = config.widgets;
export const views = config.views;
export const blocks = config.blocks;
export const addonRoutes = [...config.addonRoutes];
export const addonReducers = { ...config.addonReducers };
export const appExtras = config.appExtras;
