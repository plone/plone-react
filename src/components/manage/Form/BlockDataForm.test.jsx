import React from 'react';
import BlockDataForm, { addVariationsFieldToSchema } from './BlockDataForm';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import configureStore from 'redux-mock-store';
import config from '@plone/volto/registry';
import { Provider } from 'react-intl-redux';

const mockStore = configureStore();

beforeAll(() => {
  config.widgets = {
    id: {},
    widget: {},
    vocabulary: {},
    choices: (props) => (
      <div className="SelectWidget">{JSON.stringify(props.choices)}</div>
    ),
    type: {},
    default: () => <div className="TextWidget" />,
  };

  config.blocks.blocksConfig.testBlock = {
    id: 'testBlock',
    variations: {
      default: {
        label: 'Default',
      },
      extra: {
        label: 'Extra',
      },
    },
  };

  config.blocks.blocksConfig.testBlockVariationsList = {
    id: 'testBlockVariationsList',
    variations: [
      {
        id: 'default',
        label: 'Default',
      },
      {
        id: 'extra',
        label: 'Extra',
      },
    ],
  };
});

describe('BlockDataForm utils', () => {
  it('addVariationsFieldToSchema should add field to schema', () => {
    const testSchema = {
      fieldsets: [{ fields: [] }],
      properties: {},
    };
    const intl = { formatMessage: () => 'untitled' };
    const schema = addVariationsFieldToSchema({
      schema: testSchema,
      intl,
      variations: {},
    });
    expect(testSchema.fieldsets[0].fields).toStrictEqual(['variation']);
    expect(schema.fieldsets[0].fields).toStrictEqual(['variation']);
    expect(schema).toBe(testSchema);
  });

  it('addVariationsFieldToSchema should add only add once to schema', () => {
    const testSchema = {
      fieldsets: [{ fields: ['variation'] }],
      properties: {},
    };
    const intl = { formatMessage: () => 'untitled' };
    const schema = addVariationsFieldToSchema({
      schema: testSchema,
      intl,
      variations: {},
    });
    expect(testSchema.fieldsets[0].fields).toStrictEqual(['variation']);
    expect(schema.fieldsets[0].fields).toStrictEqual(['variation']);
    expect(schema).toBe(testSchema);
  });
});

describe('BlockDataForm', () => {
  it('should add variations to schema', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const testSchema = {
      fieldsets: [{ title: 'Default', id: 'default', fields: [] }],
      properties: {},
      required: [],
    };
    const formData = {
      '@type': 'testBlock',
    };
    const { container } = render(
      <Provider store={store}>
        <BlockDataForm formData={formData} schema={testSchema} />
      </Provider>,
    );
    expect(container).toMatchSnapshot();

    // schema is cloned, not mutated in place
    expect(testSchema.fieldsets[0].fields).toStrictEqual([]);
  });

  it('can handle variations as a list', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const testSchema = {
      fieldsets: [{ title: 'Default', id: 'default', fields: [] }],
      properties: {},
      required: [],
    };
    const formData = {
      '@type': 'testBlockVariationsList',
    };
    const { container } = render(
      <Provider store={store}>
        <BlockDataForm formData={formData} schema={testSchema} />
      </Provider>,
    );
    expect(container).toMatchSnapshot();

    // schema is cloned, not mutated in place
    expect(testSchema.fieldsets[0].fields).toStrictEqual([]);
  });
});
