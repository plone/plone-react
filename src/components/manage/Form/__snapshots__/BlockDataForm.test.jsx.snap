// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockDataForm can handle variations as a list 1`] = `
<div>
  <div
    class="ui form"
  >
    <div
      id="blockform-fieldset-default"
    >
      <div
        class="ui segment form attached"
      >
        <div
          class="SelectWidget"
        >
          [["default","Default"],["extra","Extra"]]
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BlockDataForm should add variations to schema 1`] = `
<div>
  <div
    class="ui form"
  >
    <div
      id="blockform-fieldset-default"
    >
      <div
        class="ui segment form attached"
      >
        <div
          class="SelectWidget"
        >
          [["default","Default"],["extra","Extra"]]
        </div>
      </div>
    </div>
  </div>
</div>
`;
